resources/net.go

```diff
@@ -0,0 +1,726 @@
+	// DBusNetworkdLinkInterface is the networkd dbus link interface.
+	DBusNetworkdLinkInterface = DBusNetworkdInterface + ".Link"
+	// DBusLinkPath is the DBus source path for network interface events.
+	DBusLinkPath = "/org/freedesktop/network1/link/"
+	// DBusEavesdropSprint is the format of dbus eavesdrop arguments.
+	DBusEavesdropSprint = "type='signal', interface='%s', path='%s', eavesdrop='true'"
+	// DBusFailMode is the dbus mode used for restarting the network service
+	DBusFailMode = "fail"
+
+	// NetworkdUnitFileDir is the location of networkd unit files.
+	NetworkdUnitFileDir = "/etc/systemd/network/"
+	// NetworkdUnitFileExt is the file extension for networkd unit files.
+	NetworkdUnitFileExt = ".network"
+	// NetworkdService is the name of the networkd service.
+	NetworkdService = "systemd-networkd.service"
+
+	// ActiveState is the systemd-networkd service's active state.
+	ActiveState = "ActiveState"
+	// ActiveStateActive is the active service state.
+	ActiveStateActive = "active"
+	// ActiveStateReloading is the reloading service state.
+	ActiveStateReloading = "reloading"
+
+	// OpState is the operational state of the device.
+	OpState = "OperationalState"
+	// OpStateRoutable us the routable operational state.
+	OpStateRoutable = "routable"
+	// OpStateNoCarrier is the no-carrier operational state.
+	OpStateNoCarrier = "no-carrier"
+	// OpStateOff is the off operational state.
+	OpStateOff = "off"
+)
+
+func init() {
+	RegisterResource("net", func() Res { return &NetRes{} })
+}
+
+// NetRes is an systemd-networkd resource.
+type NetRes struct {
+	BaseRes `yaml:",inline"`
+	State   string        `yaml:"state"`   // up or down
+	IPv4    string        `yaml:"ipv4"`    // ipv4 address in cidr format
+	IPv6    string        `yaml:"ipv6"`    // ipv6 address in cidr format
+	Gateway string        `yaml:"gateway"` // gateway address
+	Timeout time.Duration `yaml:"timeout"` // connection timeout in seconds
+
+	// read-only parameters
+	ActualIPv4 string // the interface's actual ipv4 address
+	ActualIPv6 string // the interface's actual ipv6 address
+
+	OpState dbus.Variant // the network link's operational state
+
+	IfacePath    dbus.ObjectPath // the dbus path to the interface
+	UnitFilePath string          // the interface unit file path
+
+	// internal busses
+	sdbus *systemd.Conn // go-systemd dbus connection
+	gdbus *dbus.Conn    // godbus dbus connection
+}
+
+// Default returns some sensible defaults for this resource.
+func (obj *NetRes) Default() Res {
+	return &NetRes{
+		BaseRes: BaseRes{
+			MetaParams: DefaultMetaParams, // force a default
+		},
+		State:   IfaceUp,
+		Timeout: time.Duration(10),
+	}
+}
+
+// Validate if the params passed in are valid data.
+func (obj *NetRes) Validate() error {
+	// validate state
+	if obj.State != IfaceUp && obj.State != IfaceDown {
+		return fmt.Errorf("invalid state: %s", obj.State)
+	}
+
+	// validate network address input
+	if obj.IPv4 != "" {
```

Comment:
I just realized that it seems the `net` package doesn't have a strong distinction between IPv4 and IPv6 (it just has an `IP` type). Do we need to add a distinction? Can we just have `obj.IP` ?